# Exercise 2: Coding Habits & Functions

This assignment will give you some practice writing your own functions and using the coding best practices discussed in the tutorial.

---

## 1. Summary statistics (4 pts)

Write a function that takes a vector of numbers `x` and returns the length, mean, and standard deviation of `x` as a new vector.

In keeping with our best practices, give the function a short but descriptive name, and use snake case if it involves multiple words.

Hint: Vectors are defined in R using the `c()` command.

In [9]:
#Write your function here
x <-4, 2, 5, ,3 


ERROR: Error in parse(text = input): <text>:2:6: unexpected ','
1: #Write your function here
2: x <-4,
        ^


In [10]:
x  <- c(2,3,4,5)

In [14]:

summary_stats <- function(x) {
    length_x  <- length(x)
    mean_x  <- mean(x)
    sd_x  <- sd(x)
    return(c(length_x, mean_x, sd_x))
}
print(paste("Summary statistics for x:",summary_stats(x)))

[1] "Summary statistics for x: 4"               
[2] "Summary statistics for x: 3.5"             
[3] "Summary statistics for x: 1.29099444873581"


Calculate the summary statistics of vector `v1`.

*Hint*: Note the "NA" in the vector. This is short for "not available" and is a placeholder for missing values. You'll want to look up the _is.na_ function (and the not operator _!_), as well as options for removing NA's in the functions that you will use to summarize the vector. Look at the documentation for the functions that you will use to see how to work with NA's.

In [15]:
v1  <- c(5, 11, 6, NA, 9)

#your_function_name(v1)

In [17]:
summary_stats(v1) # has NA's, next task is to address this 

[1]  5 NA NA

In [ ]:
#New and improved summary statistics function 

In [22]:
summary_stats  <- function (x) {
    length_x  <- length(x[!is.na(x)]) #ignores NA values uses only true values
    mean_x  <- mean(x,na.rm = TRUE) #removes NA's 
    sd_x  <- sd(x,na.rm = TRUE)
    return(c(length_x, mean_x, sd_x))
}
summary_stats(v1)
print(paste("Summary statistics for v1:",summary_stats(v1)))

[1] 4.000000 7.750000 2.753785

[1] "Summary statistics for v1: 4"               
[2] "Summary statistics for v1: 7.75"            
[3] "Summary statistics for v1: 2.75378527364305"


# ---
## 2. T-test function (4 pts)

The formula for a t-test is:

In [58]:
x <- c(2, 3, 4, 5)
ttest_fun  <- function(x, mu) {#have 2 arguments as told by the Hint 
    summary_stats  <- summary_stats(x) #length, mean, std dev
    n  <- summary_stats [1] #length
    m <- summary_stats [2] #mean
    s  <- summary_stats [3] #std_dev
    t_stat <- (m - mu)/(s/sqrt(n))
}

In [59]:
attempt_1  <- ttest_fun(x, mu = 3)
print(paste("T-test 1 =", attempt_1))

[1] "T-test 1 = 0.774596669241483"


$$ \frac{m- \mu}{ \frac{s}{\sqrt{n}}} $$

Where m is the sample mean, $\mu$ (mu) is the population mean, s is the standard deviation, and n is the sample size.

Using your function above as a starting point, write a new function `ttest_fun` that compares a vector `x` to a given population mean `mu` and calculates the t-statistic. Keep the coding best practices in mind.

Hint: You will need to add another argument for mu.


---
---
### If you use Generative AI (optional)
If you use code-assisting generative AI (e.g., Copilot, GPT, Claude; more than simple google searching), follow the steps below. If you **do not** use generative AI, skip this section and go directly to the coding part.

Your workflow should be
1. Write pseudocode
2. Generate code using generative AI
3. Compare and verify the code


**Definition**: Pseudocode is a short description of an algorithm written in plain, structured steps that a human (or an AI!) can easily read and translate into real code. Pseudocode usually does not run as. It does not need to be perfect—its purpose is to help you organize your approach and to provide clear instructions to the AI.

You may write pseudocode as comments (starting with #) inside a code cell or as plain text.

#### Step 1: Write pseudocode

Write pseudocode for `ttest_fun` below.

# Luz's pseudo code here

### define the vector I want to use
 x = 2,3,4,5

### write the function for a t-test in R
- ttest_fun = function(x, mu){
- summary stats = summary stats of x (1) # called for n of data, mean, standard deviation 
- n = summary stat (1) # first value popped out, length or n
- m = summary stat (2) # second value popped out, mean
- s = summary stat (3) # third value popped out, standard deviation 
- t_stat = (m-mu)/(s/square root of (n)) #t-test formula }

#### Step 2: Generate code using generative AI
Using your pseudocode, prompt a generative AI to create or fix the R functionso it follows the instructions. Then paste **only the generated function** below (not the full AI response).

### define the vector I want to use
x <- c(2, 3, 4, 5)

### write the function for a t-test in R
ttest_fun <- function(x, mu) {
  summary_stats <- calc_summary_stats(x)  # called for n of data, mean, standard deviation 
  n <- summary_stats[1]  # first value popped out, length or n
  m <- summary_stats[2]  # second value popped out, mean
  s <- summary_stats[3]  # third value popped out, standard deviation 
  t_stat <- (m - mu) / (s / sqrt(n))  # t-test formula
  
  return(t_stat)
    # Example: test against population mean of 3
result <- ttest_fun(x, mu = 3)
print(result)
# Output: 0.7745967

}


#### Step 3: Compare and verify the code

Compare your pseudocode and the AI-generated code. Answer the questions below concisely (a few bullet points is enough.)

**Question 1.** What is different between your pseudocode and generated code.
(e.g., it handled NAs differently, used a different formula arrangement, etc.)

**Question 2.** What are the input parameters and the output of the function?
(Answer to show you understand what the function takes in and returns. Be concise.)
- 
**Question 3.** Does anything has to be changed? (Yes / No)

Write your answer here
##### Question 1.
- There are different names used and it included a return at the end while I had done a print(paste). It also tested the function and defined mu as 3 during the function use and printed the result then. 

##### Question 2.
- Inputs are the arithmetic functions needed to calculate a t-test of n of the group, mean, and standard deviation. The group here is the vector we choose to use (in this notebook its x or v1). Output is the single value of the t-test
##### Question 3.
-  No

If you answers yes to the question 3, revise the code (either using GenAI or on your own). Don't forget to paste your final version in the code block below.

---
---

In [ ]:
# write ttest_fun here


Use your function to compare the mean of v1 to 10.

In [62]:
ttest_fun(v1, 10)

result  <- ttest_fun(v1, 10)
print(result)

[1] -1.634114


---
## 3. Setting default values (2 pts)

Set the default value of mu to 0. Test your modified function below by supplying only `v2` as an argument.

In [63]:
# Write your modified ttest_fun here


In [66]:
v2 <- c(3, 7, 1, NA, 8, 12)
result_2 <- ttest_fun(v2, mu = 0)
print(result_2)

[1] 3.205944


How does your result compare to R's built-in `t.test()` function?

In [67]:
t.test(v2)


	One Sample t-test

data:  v2
t = 3.2059, df = 4, p-value = 0.03272
alternative hypothesis: true mean is not equal to 0
95 percent confidence interval:
  0.8306107 11.5693893
sample estimates:
mean of x 
      6.2 


- My functions result the t.test function result is the same but it also offers more details of degrees of freedom, p-val, confidence interval, mean, and alt hypothesis. Also it was way simpler to use the in

When you are finished, save the notebook as Exercise2.ipynb, push it to your class GitHub repository (the one you made for Exercise 1) and send the instructors a link to your notebook via Canvas. You can send messages via Canvas by clicking "Inbox" on the left and then pressing the icon with a pencil inside a square.

**DUE:** 5pm EST, Jan 29, 2026

**IMPORTANT** Did you collaborate with anyone on this assignment? If so, list their names here.
> *Someone's Name*